src/search/pdbs/pattern_collection_generator_multiple.cc

```diff
@@ -0,0 +1,310 @@
+            if (verbosity >= utils::Verbosity::NORMAL) {
+                utils::g_log << "stagnation limit reached, terminating" << endl;
+            }
+            return true;
+        }
+    }
+    return false;
+}
+
+static PatternCollectionInformation get_pattern_collection(
+    const TaskProxy &task_proxy, const shared_ptr<PDBCollection> &pdbs) {
+    shared_ptr<PatternCollection> patterns = make_shared<PatternCollection>();
+    patterns->reserve(pdbs->size());
+    for (const shared_ptr<PatternDatabase> &pdb : *pdbs) {
+        patterns->push_back(pdb->get_pattern());
+    }
+    PatternCollectionInformation result(task_proxy, patterns);
+    result.set_pdbs(pdbs);
+    return result;
+}
+
+PatternCollectionInformation PatternCollectionGeneratorMultiple::generate(
+    const shared_ptr<AbstractTask> &task) {
+    if (verbosity >= utils::Verbosity::NORMAL) {
+        utils::g_log << "Generating patterns using the " << get_name()
+                     << " algorithm." << endl;
+    }
+
+    TaskProxy task_proxy(*task);
+    utils::CountdownTimer timer(total_max_time);
+
+    // Store the set of goals in random order.
+    vector<FactPair> goals = get_goals_in_random_order(task_proxy, *rng);
+
+    // Store the non-goal variables for potential blacklisting.
+    vector<int> non_goal_variables = get_non_goal_variables(task_proxy);
+
+    if (verbosity >= utils::Verbosity::DEBUG) {
+        utils::g_log << "goal variables: ";
+        for (FactPair goal : goals) {
+            utils::g_log << goal.var << ", ";
+        }
+        utils::g_log << endl;
+        utils::g_log << "non-goal variables: " << non_goal_variables << endl;
+    }
+
+    initialize(task);
+
+    // Collect all unique patterns and their PDBs.
+    set<Pattern> generated_patterns;
+    shared_ptr<PDBCollection> generated_pdbs = make_shared<PDBCollection>();
+
+    shared_ptr<utils::RandomNumberGenerator> pattern_computation_rng =
+        make_shared<utils::RandomNumberGenerator>(random_seed);
+    int num_iterations = 1;
+    int goal_index = 0;
+    /*
+      Start blacklisting after the percentage of max_time specified via
+      blacklisting_trigger_percentage has passed. Compute this time point once.
+    */
+    double blacklisting_start_time = total_max_time * blacklist_trigger_percentage;
+    while (true) {
+        check_blacklist_trigger_timer(blacklisting_start_time, timer);
+
+        unordered_set<int> blacklisted_variables =
+            get_blacklisted_variables(non_goal_variables);
+
+        int remaining_pdb_size = min(remaining_collection_size, max_pdb_size);
+        double remaining_time =
+            min(static_cast<double>(timer.get_remaining_time()), pattern_generation_max_time);
+
+        PatternInformation pattern_info = compute_pattern(
+            remaining_pdb_size,
+            remaining_time,
+            pattern_computation_rng,
+            task,
+            goals[goal_index],
+            move(blacklisted_variables));
```

Comment:
I didn't look at how this is used yet, but it seems strange that we move `blacklisted_variables` instead of passing a const reference.

Reply:
I agree, but we left this like this because CEGAR not only wants to store this set of variables, but also extends it.